app/uk/gov/hmrc/teamsandrepositories/connectors/BuildDeployApiConnector.scala

```
@@ -0,0 +1,139 @@
+                  )
+      _        <- if (response.success)
+                    Future.unit
+                  else
+                    Future.failed(new Throwable(s"Failed to set branch protection for $repoName: ${response.message}"))
+    } yield ()
+  }
+
+  private lazy val signer =
+    new AWSSigner(
+      awsCredentialsProvider,
+      config.awsRegion,
+      "execute-api"
+    )
+}
+
+object BuildDeployApiConnector {
+
+  final case class Request(
+    repoName: String,
+    enable: Boolean
+  )
+
+  object Request {
+
+    implicit val writes: Writes[Request] =
+      ( (__ \ "repository_names"          ).write[String]
+      ~ (__ \ "set_branch_protection_rule").write[Boolean]
+      )(unlift(Request.unapply))
+  }
+
+  final case class Response(
+    success: Boolean,
+    message: String
+  )
+
+  object Response {
+
+    implicit val reads: Reads[Response] =
+      Reads.list {
+        ( (__ \ "success").read[Boolean]
+        ~ (__ \ "message").read[String]
+        ) (Response.apply _)
+      }.collect(JsonValidationError("A single-element array of responses was expected but not found")) {
+        case x :: Nil => x
+      }
+  }
+
+  final class AWSSigner(
+    awsCredentials: AWSCredentialsProvider,
+    awsRegion: String,
+    awsService: String
+  ) {
+
+    def sign(
+      url: URL,
+      host: Option[String] = None,
+      method: String,
+      payload: Option[JsValue] = None
+    ): Seq[(String, String)] = {
+      AwsSigner(awsCredentials, awsRegion, awsService, () => LocalDateTime.now())
+        .getSignedHeaders(
+          uri = url.getPath,
+          method = method,
+          queryParams = toMap(url.getQuery),
+          headers = Map[String, String]("host" -> host.getOrElse(url.getHost)),
+          payload = payload.map(Json.toBytes)
+        ).toSeq
+    }
+
+    // converts a query string into a map, note this would fall over if the same parameter was used twice
```

can we fail fast if this occurs?